bite-size-notes/week-3-a-cbf-jingyi.md

```diff
@@ -0,0 +1,70 @@
+| Trade-offs              | May be conservative in their safety guarantees                  | Require stronger assumptions about the system dynamics and stability properties  |
+| Complementary Nature    | Focus on safety-critical aspects, ensuring the system remains within safe operating bounds | Address stability concerns, guaranteeing convergence to desired equilibrium points  |
+## Combine CBF Constraints with CLF Constraints
+#### Motivation
+In many practical scenarios, we may want to design a system that aims to reach a desired goal while simultaneously avoiding obstacles. This can be achieved by combining Control Lyapunov Function (CLF) constraints and Control Barrier Function (CBF) constraints.
+#### Stpes
+To balance the objective of reaching a goal and avoiding obstacles, we can formulate an optimization problem with the following steps:
+1. Choose a desired control input $u^{dsr}$ that is as close as possible to the actual control input $u$. This can be achieved by minimizing the squared norm $\underset{u}\min \lVert u^{dsr}-u \rVert^2_2$, subject to the constraint $u \in \mathcal{U}$.
+2. Include the CLF constraint $\nabla V(x)^T f(x,u) \le -\beta V(x)$, where $V(x)$ is a Control Lyapunov Function and $\beta$ is a positive parameter. This ensures that the system is making progress towards the desired goal.
+3. Include the CBF constraint $\nabla b(x)^T f(x,u) \ge -\alpha(b(x))$, where $b(x)$ is the Control Barrier Function and $\alpha$ is a class $\mathcal{K}$ function. This ensures that the system stays within the safe set defined by $b(x) \le 0$.
+4. However, combining the CLF and CBF constraints can lead to an infeasible optimization problem. To address this, we can introduce a slack variable $\epsilon \ge 0$ and relax the CLF constraint to $\nabla V(x)^T f(x,u) \le \epsilon$. This allows the system to slightly deviate from the desired goal for the sake of safety.
+5. To minimize the violation of the CLF constraint, we can add a penalty term $\gamma\epsilon^2$ to the objective function, where $\gamma$ is a weighting factor. The final optimization problem becomes: $\underset{u, \epsilon}\min \lVert u^{dsr}-u \rVert^2_2 + \gamma \epsilon^2$ subject to:
+    - $u \in \mathcal{U}$
+    - $\nabla V(x)^T f(x,u) \le \epsilon$
+    - $\nabla b(x)^T f(x,u) \ge -\alpha(b(x))$
+#### Conclusion
+The resulting optimization problem is a Quadratic Program (QP) because the objective function is quadratic, and the constraints are linear in the control input $u$ and the slack variable $\epsilon$. This formulation allows the system to balance the competing objectives of reaching the desired goal and avoiding obstacles.
+- Additional Notes
+    - If the system dynamics are control-affine, i.e., $f(x,u) = f_0(x) + f_1(x)u$, then the optimization problem becomes linear in the control input $u$.
+## Conclusion
```

Comment:
This conclusion is a sum up for "Combine CBF Constraints with CLF Constraints" section. But I can see the confusion. Since the part is quite necessary, I have replaced the subtitle to avoid confusion.